### ISSUE TITLE
[Bug/Support]: git-commit blocked on magit, and magit blocked on git-commit: circular dep?

### ISSUE THREAD

**Issue by garyo** [progfolio/elpaca#302]
### Confirmation

- [X] I have checked the documentation (README, Wiki, docstrings, etc)
- [ ] I am  checking these without reading them.
- [X] I have searched previous issues to see if my question is a duplicate.

### Elpaca Version

Elpaca 43ec2d8 grafted, HEAD -> master, origin/master, origin/HEAD
installer:      0.7
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-w64-mingw32)
 of 2024-04-08
git --version:  git version 2.44.0.windows.1

### Operating System

Windows 11

### Description

Doing `elpaca-update-all` on my Windows machine, with recent Emacs 30 and latest elpaca, I have what seems like a circular dependency that prevents elpaca from finishing: magit depends on git-commit, and git-commit is inside the magit repo so is blocked on magit. I'm not sure how to proceed.

```
git-commit                     blocked              Waiting for mono-repo magit                                                      27.152611
magit-section                  blocked              Waiting for mono-repo magit                                                      27.152635
magit                          blocked              Blocked by: (git-commit) 
```

(I'm also having unrelated issues with "too many open files" errors, still working on those, but wanted to report this one.)

**Comment by progfolio** [comment]
Thanks for taking the time to fill out a support ticket. 

> Doing `elpaca-update-all` on my Windows machine, with recent Emacs 30 and latest
> elpaca, I have what seems like a circular dependency that prevents elpaca from
> finishing: magit depends on git-commit, and git-commit is inside the magit repo
> so is blocked on magit. I'm not sure how to proceed.
> 
> ```
> git-commit                     blocked              Waiting for mono-repo magit                                                      27.152611
> magit-section                  blocked              Waiting for mono-repo magit                                                      27.152635
> magit                          blocked              Blocked by: (git-commit) 
> ```

Yes. That does look like a circular dependency.
I've squashed that bug several times, so perhaps there's a case I missed.

> (I'm also having unrelated issues with "too many open files" errors, still
> working on those, but wanted to report this one.)

Maybe they're related.
You can prevent the "too many open file" error by setting the `elpaca-queue-limit` option prior to processing any queues. e.g.

``` emacs-lisp
(setq elpaca-queue-limit 30)
```

If you search the issue tracker there are other Windows users who have hit this limit.
I don't recall exactly where they ended up with that value, but I think it was closer to 12-20. Allegedly there are ways to allow more open file handles at the OS level, but I don't use Windows enough to give any advice on how to do it.

I would try:

1. saving `(setq elpaca-queue-limit 12)` in you init file just after the elpaca installer.
2. `M-x restart-emacs`

Then, in a fresh Emacs session, try another `elpaca-update-all` and see if the issue persists.

**Comment by garyo** [comment]
This specific issue seems to be gone now since I applied a user fix (see #303) but I'm still getting an issue with transient; magit-commit requires a very recent version of transient (>=20240421) but elpaca doesn't install that, so the magit-commit install fails.
The log is:
```
git-commit [MELPA|NonGNU-devel ELPA]
Edit Git commit messages.

source: MELPA
url: https://github.com/magit/magit
menu item recipe:
( :package "git-commit"
 :fetcher github
 :repo "magit/magit"
 :files ("lisp/git-commit.el" "lisp/git-commit-pkg.el")
 :old-names (git-commit-mode)
 :source "MELPA")
full recipe:
( :package "git-commit" 
 ;; Inherited from elpaca-order-functions.
 :depth 1
 :inherit t
 :protocol https
 ;; Inherited from elpaca-menu-item.
 :source "MELPA"
 :old-names (git-commit-mode)
 :files ("lisp/git-commit.el" "lisp/git-commit-pkg.el")
 :repo "magit/magit"
 :fetcher github)
dependencies:
  emacs       >= 26.1
  compat      >= 29.1.4.5
  transient   >= 20240421
  with-editor >= 20240415
dependents: 
  magit
installed version: 3.3.0.50-git f7cba11
statuses:
  (failed unblocked blocked continued-dep queued)
files:
  $REPOS/magit/lisp/git-commit-pkg.el ! $BUILDS/git-commit/git-commit-pkg.el
  $REPOS/magit/lisp/git-commit.el     ! $BUILDS/git-commit/git-commit.el
log:
  [2024-05-08 12:20:41] Package queued
  [2024-05-08 12:20:41] Continued by: elpaca--continue-dependency
  [2024-05-08 12:20:41] Continued by: elpaca--dispatch-build-commands
  [2024-05-08 12:20:41] Queueing Dependencies
  [2024-05-08 12:20:44] Continued by: elpaca--check-status
  [2024-05-08 12:20:44] Unblocked by: with-editor
  [2024-05-08 12:20:44] Checking dependency versions
  [2024-05-08 12:20:44] transient installed version (20240408) lower than min required 20240421
  [2024-05-08 12:20:44] Continued by: elpaca--check-version
```

If I do `(use-package transient)` before `(use-package magit)` it works, but I don't think I should need to do that -- maybe it's due to the older version of transient being built-in, so elpaca doesn't want to silently override it on my behalf?

**Comment by progfolio** [comment]
> This specific issue seems to be gone now since I applied a user fix (see #303)

Glad to hear it.

> If I do (use-package transient) before (use-package magit) it works, but I don't think I should need to do that -- maybe it's due to the older version of transient being built-in, so elpaca doesn't want to silently override it on my behalf?

You're correct. Currently, Elpaca will not install a built-in package unless you explicitly request it.
There's a feature request for automatically updating built-in packages and some previous discussion here: https://github.com/progfolio/elpaca/issues/272

> but I'm still getting an issue with transient; magit-commit requires a very recent version of transient (>=20240421) but elpaca doesn't install that, so the magit-commit install fails.

What's specifically happening in this situation is:

git-commit requires transient with a "YYYYMMDD" timestamp version (which MELPA introduced, and very few package authors use to declare dependencies).
Since transient is built-in, we have no reliable way of getting the exact date it was last modified.
Elpaca's compromise is to maintain an alist of Emacs release dates for stable versions and fall back to `emacs-build-time` for unstable versions when available.
That value is stored in `elpaca-core-date` which is used as the timestamp version for any built-in package.
My hunch is that your `elpaca-core-date` is "20240408" considering:

```
emacs-version: GNU Emacs 30.0.50 (build 1, x86_64-w64-mingw32) of 2024-04-08
```

Rebuilding Emacs should bump that value and transient will pass the version check.
It's also possible to work around the issue by lying to Elpaca about the version date (via the `:version` recipe keyword), or disabling the check altogether, but that may lead to breakage.
